src/main/java/com/google/cloud/tools/appengine/cloudsdk/serialization/CloudSdkVersion.java

```diff
@@ -20,71 +20,119 @@
+  }
+
+  private static String getSemVerRegex() {
+    // Only digits, with no leading zeros.
+    String digits = "(?:0|[1-9][0-9]*)";
+    // Digits, letters and dashes
+    String alphaNum = "[-0-9A-Za-z]+";
+    // This is an alphanumeric string that must have at least one letter (or else it would be
+    // considered digits).
+    String strictAlphaNum = "[-0-9A-Za-z]*[-A-Za-z]+[-0-9A-Za-z]*";
+
+    String preReleaseIdentifier = "(?:" + digits + "|" + strictAlphaNum + ")";
+    String preRelease = "(?:" + preReleaseIdentifier + "(?:\\." + preReleaseIdentifier + ")*)";
+    String build = "(?:" + alphaNum + "(?:\\." + alphaNum + ")*)";
+
+    return "^(?<major>" + digits + ")\\.(?<minor>" + digits + ")\\.(?<patch>" + digits + ")"
+        + "(?:\\-(?<prerelease>" + preRelease + "))?(?:\\+(?<build>" + build + "))?$";
   }
 
   @Override
   public String toString() {
     return version;
   }
 
+  /**
+   * Compares this to another CloudSdkVersion, per the Semantic Versioning 2.0.0 specification. Note
+   * that the build identifier field is excluded for comparison.
+   */
   @Override
-  public int compareTo(CloudSdkVersion otherVersion) {
-    List<Integer> mine = new ArrayList<>(this.versionComponents);
-    List<Integer> other = new ArrayList<>(otherVersion.getVersionComponents());
-
-    // if both versions don't have the same number of components, pad the smaller one with zeros
-    rightPadZerosUntilSameLength(mine, other);
+  public int compareTo(CloudSdkVersion other) {
+    Preconditions.checkNotNull(other);
 
-    // compare version integers from left to right
+    // First, compare required fields
+    List<Integer> mine = ImmutableList.of(majorVersion, minorVerion, patchVersion);
+    List<Integer> others = ImmutableList.of(other.getMajorVersion(), other.getMinorVerion(),
+        other.getPatchVersion());
     for (int i = 0; i < mine.size(); i++) {
-      int result = mine.get(i).compareTo(other.get(i));
+      int result = mine.get(i).compareTo(others.get(i));
       if (result != 0) {
         return result;
       }
     }
+
+    // Compare pre-release components
+    if (preRelease != null && other.getPreRelease() != null) {
+      return preRelease.compareTo(other.getPreRelease());
+    }
+
+    // A SemVer with a pre-release string has lower precedence than one without.
+    if (preRelease == null && other.getPreRelease() != null) {
+      return 1;
+    }
+    if (preRelease != null && other.getPreRelease() == null) {
+      return -1;
+    }
+
     return 0;
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(versionComponents);
+    return Objects.hash(majorVersion, minorVerion, patchVersion, preRelease);
   }
 
   /**
-   * Compares objects for equality. Two CloudSdkVersions are considered equal if, from left to
-   * right, their version integers are equal.
+   * Compares this to another CloudSdkVersion for equality. Note that the build identifier field is
+   * excluded for comparison.
```

Comment:
The confusion was mine wrgt compareTo() signature. I didn't notice github had folded everything before the return statement.

WRGT equals vs compareto,  I don't think that it's a requirement of the compareTo contract that compareTo == 0 always implies the two objects are equal.   It's odd to have two objects be equals with different toString values and/or responses to getters.